converter/internal/otelcolconvert/otelcolconvert.go

```diff
@@ -0,0 +1,255 @@
+//
+// For compatibility with other converters, the extraArgs paramater is defined
+// but unused, and a critical error diagnostic is returned if extraArgs is
+// non-empty.
+func Convert(in []byte, extraArgs []string) ([]byte, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	if len(extraArgs) > 0 {
+		diags.Add(diag.SeverityLevelCritical, fmt.Sprintf("extra arguments are not supported for the otelcol converter: %s", extraArgs))
+		return nil, diags
+	}
+
+	cfg, err := readOpentelemetryConfig(in)
+	if err != nil {
+		diags.Add(diag.SeverityLevelCritical, err.Error())
+		return nil, diags
+	}
+	if err := cfg.Validate(); err != nil {
+		diags.Add(diag.SeverityLevelError, fmt.Sprintf("failed to validate config: %s", err))
+		return nil, diags
+	}
+
+	f := builder.NewFile()
+
+	diags.AddAll(appendConfig(f, cfg))
+	diags.AddAll(common.ValidateNodes(f))
+
+	var buf bytes.Buffer
+	if _, err := f.WriteTo(&buf); err != nil {
+		diags.Add(diag.SeverityLevelCritical, fmt.Sprintf("failed to render Flow config: %s", err.Error()))
+		return nil, diags
+	}
+
+	if len(buf.Bytes()) == 0 {
+		return nil, diags
+	}
+
+	prettyByte, newDiags := common.PrettyPrint(buf.Bytes())
+	diags.AddAll(newDiags)
+	return prettyByte, diags
+}
+
+func readOpentelemetryConfig(in []byte) (*otelcol.Config, error) {
+	provider := yamlprovider.New()
+
+	configProvider, err := otelcol.NewConfigProvider(otelcol.ConfigProviderSettings{
+		ResolverSettings: confmap.ResolverSettings{
+			URIs: []string{"yaml:" + string(in)},
+			Providers: map[string]confmap.Provider{
+				provider.Scheme(): provider,
+			},
+			Converters: []confmap.Converter{expandconverter.New()},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create otelcol config provider: %w", err)
+	}
+
+	cfg, err := configProvider.Get(context.Background(), getFactories())
+	if err != nil {
+		// TODO(rfratto): users may pass unknown components in YAML here. Can we
+		// improve the errors? Can we ignore the errors?
+		return nil, fmt.Errorf("failed to get otelcol config: %w", err)
+	}
+
+	return cfg, nil
+}
+
+func getFactories() otelcol.Factories {
+	facts := otelcol.Factories{
+		Receivers:  make(map[component.Type]receiver.Factory),
+		Processors: make(map[component.Type]processor.Factory),
+		Exporters:  make(map[component.Type]exporter.Factory),
+		Extensions: make(map[component.Type]extension.Factory),
+		Connectors: make(map[component.Type]connector.Factory),
+	}
+
+	for _, converter := range converters {
+		fact := converter.Factory()
+
+		switch fact := fact.(type) {
+		case receiver.Factory:
+			facts.Receivers[fact.Type()] = fact
+		case processor.Factory:
+			facts.Processors[fact.Type()] = fact
+		case exporter.Factory:
+			facts.Exporters[fact.Type()] = fact
+		case extension.Factory:
+			facts.Extensions[fact.Type()] = fact
+		case connector.Factory:
+			facts.Connectors[fact.Type()] = fact
+
+		default:
+			panic(fmt.Sprintf("unknown component factory type %T", fact))
```

Comment:
WDYT about a diag on this rather than a panic? This seems like an unsupported ERROR diag ideally where the rest of the config can be potentially converted. If not possible to continue conversion here for otel, maybe a CRITICAL?

Reply:
There are some other panics throughout I think it is worth a pass to see if you think they can become diags or not.

Reply:
Generally speaking, I use panics to indicate that the code is fundamentally broken; users shouldn't be able to run into these things. I can change these out with diags, but I also think that comes at a cost of complicating the code, as now I have to wire in more `diags.AddAll` calls and also check to see if a critical error has been generated yet. 

I see a few options here:

* Replace panics with critical diagnostics as you suggested
* Catch panics at the call to `Convert` and turn them into critical diagnostics to avoid complicating code 
* Leave the panics as-is with no changes

What would you prefer? The converters are your code so I'd rather follow your lead here

Reply:
I don't think this is a blocking thing if it complicates the code. We can always come back to it later if it becomes a stumbling point for users. Catching panics is interesting but I'm worried it will catch other panics and not display in a useful way for debugging if we start catching them and not display the whole error stack which kind of defeats the purpose.